Implements arch type functions & refactors out "_settings" global variable.

skl131313 opened the pull request:
Also requires the code-d commit to match. There was a global _settings variable that was being passed around. From the looks of it, it was possible to simply keep adding and adding flags to it without it ever clearing itself anywhere. The dub function "determinePlatform" that takes it in to add flags never clears it either. So I simply removed it as it didn't seem to be needed anyways.

Commit:
Implements arch type functions & refactors out "_settings" global variable.

WebFreak001 commented on source/com/dub.d:
```diff
@@ -210,6 +210,41 @@ bool setConfiguration(string configuration)
+@arguments("subcmd", "list:arch-types")
+string[] archTypes() @property
+{
+	return [ "x86", "x86_64" ]; // todo update this list to include all other archs
```
gdc also has `arm` and `arm_thumb`, can you also add those when `compilerName.toLower.contains("gdc")`?

https://github.com/dlang/dub/blob/8f87d9f8413d9e962331accd26d7ccee9ab69b35/source/dub/compilers/gdc.d#L70-L73

WebFreak001 replied:
Or maybe using the getCompiler method would be better

`if (cast(GDCCompiler) getCompiler(compilerName)) return ["x86", "x86_64", "arm", "arm_thumb"];`

https://github.com/dlang/dub/blob/8f87d9f8413d9e962331accd26d7ccee9ab69b35/source/dub/compilers/compiler.d#L32-L44

Commit:
Adds more arch types specific for gdc.

Commit:
Fixes issue where config name used in place of actual binary file name.

skl131313 commented:
I added them but the version of gdc I was using didn't actually support them as it was missing the -marm parameter.

Also added a fix for the compiler name being used. Eg when "ldc2" was set, it would use Dub's "interface Compiler.name()" function which isn't the actual binary name of the compiler that should be used. So if set to "ldc2" it would then be changed to "ldc", meaning it would fail to compile.

Commit:
Removes global vars _compiler and _platform.
